ratis-server/src/main/java/org/apache/ratis/server/leader/LogAppenderBase.java

```diff
@@ -225,4 +230,14 @@ public InstallSnapshotRequestProto newInstallSnapshotNotificationRequest(TermInd
   public Iterable<InstallSnapshotRequestProto> newInstallSnapshotRequests(String requestId, SnapshotInfo snapshot) {
     return new InstallSnapshotRequests(server, getFollowerId(), requestId, snapshot, snapshotChunkMaxSize);
   }
+
+  @Override
+  public void registerAppendEntriesListener(BiConsumer<RaftPeerId, AppendEntriesReplyProto> listener) {
+    appendEntriesListener = listener;
+  }
+
+  protected void notifyAppendEntriesListener(AppendEntriesReplyProto reply) {
+    Optional.ofNullable(appendEntriesListener).ifPresent(
+        listener -> listener.accept(this.getServer().getId(), reply));
```

Comment:
Let's add a new method to `LeaderState`
```java
//LeaderState.java
  /** Received an {@link AppendEntriesReplyProto} */
  void onAppendEntriesReply(FollowerInfo follower, AppendEntriesReplyProto reply);
```
Then, this method becomes:
```java
    getLeaderState().onAppendEntriesReply(getFollower(), reply);
```

Reply:
done